File: app/desktop/src/main/services/vcfg.ts
import * as fs from "fs";
import * as path from "path";
import type { VcfgStateSummary, VcfgSnapshot } from "../../renderer/types";

interface VdfNode {
  [key: string]: string | VdfNode;
}


function tokenize(content: string): string[] {
  const tokens: string[] = [];
  let index = 0;

  while (index < content.length) {
    const char = content[index];
    if (/\s/.test(char)) {
      index++;
      continue;
    }
    if (char === "/" && content[index + 1] === "/") {
      index = content.indexOf("\n", index + 2);
      if (index === -1) break;
      continue;
    }
    if (char === "{" || char === "}") {
      tokens.push(char);
      index++;
      continue;
    }
    if (char !== '"') {
      index++;
      continue;
    }

    index++;
    let value = "";
    while (index < content.length) {
      const current = content[index++];
      // CS2's generated KeyValues uses a literal backslash key as "\";
      // do not apply JSON/C-style escape handling here.
      if (current === '"') {
        break;
      } else {
        value += current;
      }
    }
    tokens.push(value);
  }

  return tokens;
}

function parseNode(tokens: string[], cursor: { value: number }): VdfNode {
  const node: VdfNode = {};
  while (cursor.value < tokens.length) {
    const key = tokens[cursor.value++];
    if (key === "}") break;
    const value = tokens[cursor.value++];
    if (value === "{") node[key] = parseNode(tokens, cursor);
    else if (value !== undefined) node[key] = value;
  }
  return node;
}

function parseVcfg(filePath: string): VdfNode | null {
  if (!fs.existsSync(filePath)) return null;
  try {
    const tokens = tokenize(fs.readFileSync(filePath, "utf-8"));
    const cursor = { value: 0 };
    const rootKey = tokens[cursor.value++];
    if (!rootKey || tokens[cursor.value++] !== "{") return null;
    return { [rootKey]: parseNode(tokens, cursor) };
  } catch {
    return null;
  }
}

function child(node: VdfNode | null, ...keys: string[]): VdfNode | null {
  let current: string | VdfNode | undefined = node ?? undefined;
  for (const key of keys) {
    if (!current || typeof current === "string") return null;
    current = current[key];
  }
  return current && typeof current !== "string" ? current : null;
}

function countEntries(node: VdfNode | null): number {
  return node ? Object.keys(node).length : 0;
}

function stringEntries(node: VdfNode | null): Record<string, string> {
  if (!node) return {};
  return Object.fromEntries(
    Object.entries(node).filter((entry): entry is [string, string] => typeof entry[1] === "string"),
  );
}

export function captureVcfgSnapshot(userCfgPath: string): VcfgSnapshot {
  const keys = parseVcfg(path.join(userCfgPath, "cs2_user_keys_0_slot0.vcfg"));
  const userConvars = parseVcfg(path.join(userCfgPath, "cs2_user_convars_0_slot0.vcfg"));
  const machineConvars = parseVcfg(path.join(userCfgPath, "cs2_machine_convars.vcfg"));

  return {
    schemaVersion: 1,
    capturedAt: Date.now(),
    userCfgPath,
    bindings: stringEntries(child(keys, "config", "bindings")),
    analogBindings: stringEntries(child(keys, "config", "analogbindings")),
    userConvars: stringEntries(child(userConvars, "config", "convars")),
    machineConvars: stringEntries(child(machineConvars, "config", "convars")),
  };
}

export interface SnapshotToCfgOptions {
  bindings: boolean;
  analogBindings: boolean;
  userConvars: boolean;
  machineConvars: boolean;
}

/** Parse a .cfg file and extract convar → value pairs, skipping comments/echo/exec/bind. */
export function parseCfgConvars(filePath: string): Record<string, string> {
  if (!fs.existsSync(filePath)) return {};
  try {
    const lines = fs.readFileSync(filePath, "utf-8").split(/\r?\n/);
    const result: Record<string, string> = {};
    for (const raw of lines) {
      const commentIdx = raw.indexOf("//");
      const code = (commentIdx >= 0 ? raw.slice(0, commentIdx) : raw).trim();
      if (!code) continue;
      if (/^(?:echo|exec|bind|binddefaults|firstperson)\b/i.test(code)) continue;
      const match = code.match(/^(\S+)\s+(?:"([^"]*)"|(\S+))$/);
      if (match) result[match[1]] = match[2] ?? match[3];
    }
    return result;
  } catch {
    return {};
  }
}

/** Normalize boolean literals so "true"/"1" and "false"/"0" compare equal. */
function normalizeCfgValue(v: string): string {
  if (v === "true" || v === "True") return "1";
  if (v === "false" || v === "False") return "0";
  return v;
}

/** Convert a VcfgSnapshot into CS2 CFG text lines, filtered by category.
 *  Convars are diffed against the optional Valve baseline; only non-default values are emitted. */
export function snapshotToCfg(
  snapshot: VcfgSnapshot,
  options: SnapshotToCfgOptions,
  baseline: Record<string, string> = {},
): string {
  const sections: string[] = [];

  if (options.bindings) {
    const entries = Object.entries(snapshot.bindings).sort(([a], [b]) => a.localeCompare(b));
    if (entries.length) {
      sections.push("// ── 按键绑定 ──");
      for (const [key, cmd] of entries) sections.push(`bind "${key}" "${cmd}"`);
    }
  }

  if (options.analogBindings) {
    const entries = Object.entries(snapshot.analogBindings).sort(([a], [b]) => a.localeCompare(b));
    if (entries.length) {
      sections.push("// ── 模拟轴绑定 ──");
      for (const [axis, cmd] of entries) sections.push(`bind "${axis}" "${cmd}"`);
    }
  }

  if (options.userConvars) {
    const entries = Object.entries(snapshot.userConvars)
      .sort(([a], [b]) => a.localeCompare(b))
      .filter(([name, value]) => {
        const def = baseline[name];
        return def === undefined || normalizeCfgValue(def) !== normalizeCfgValue(value);
      });
    if (entries.length) {
      sections.push("// ── 个人偏好设置（仅与 Valve 默认值不同的项）──");
      for (const [convar, value] of entries) {
        sections.push(value.includes(" ") ? `${convar} "${value}"` : `${convar} ${value}`);
      }
    }
  }

  if (options.machineConvars) {
    const entries = Object.entries(snapshot.machineConvars)
      .sort(([a], [b]) => a.localeCompare(b))
      .filter(([name, value]) => {
        const def = baseline[name];
        return def === undefined || normalizeCfgValue(def) !== normalizeCfgValue(value);
      });
    if (entries.length) {
      sections.push("// ── 机器设置（仅与 Valve 默认值不同的项）──");
      for (const [convar, value] of entries) {
        sections.push(value.includes(" ") ? `${convar} "${value}"` : `${convar} ${value}`);
      }
    }
  }

  return sections.join("\n");
}

export function saveVcfgBaseline(
  userCfgPath: string,
  snapshotRoot: string,
  accountId: string,
): { path: string; created: boolean } {
  const accountDir = path.join(snapshotRoot, accountId);
  const baselinePath = path.join(accountDir, "baseline.json");
  if (fs.existsSync(baselinePath)) return { path: baselinePath, created: false };

  fs.mkdirSync(accountDir, { recursive: true });
  fs.writeFileSync(
    baselinePath,
    JSON.stringify(captureVcfgSnapshot(userCfgPath), null, 2),
    "utf-8",
  );
  return { path: baselinePath, created: true };
}

export function inspectVcfgState(userCfgPath: string | null): VcfgStateSummary {
  if (!userCfgPath) {
    return {
      available: false,
      bindings: 0,
      analogBindings: 0,
      cloudConvars: 0,
      machineConvars: 0,
      hasCloudMirror: false,
      hasVideoConfig: false,
    };
  }

  const keysPath = path.join(userCfgPath, "cs2_user_keys_0_slot0.vcfg");
  const userConvarsPath = path.join(userCfgPath, "cs2_user_convars_0_slot0.vcfg");
  const machineConvarsPath = path.join(userCfgPath, "cs2_machine_convars.vcfg");
  const keys = parseVcfg(keysPath);
  const userConvars = parseVcfg(userConvarsPath);
  const machineConvars = parseVcfg(machineConvarsPath);
  const bindings = stringEntries(child(keys, "config", "bindings"));

  return {
    available: Boolean(keys || userConvars || machineConvars),
    bindings: Object.keys(bindings).length,
    analogBindings: countEntries(child(keys, "config", "analogbindings")),
    cloudConvars: countEntries(child(userConvars, "config", "convars")),
    machineConvars: countEntries(child(machineConvars, "config", "convars")),
    hasCloudMirror:
      fs.existsSync(`${keysPath}_lastclouded`) ||
      fs.existsSync(`${userConvarsPath}_lastclouded`),
    hasVideoConfig: fs.existsSync(path.join(userCfgPath, "cs2_video.txt")),
  };
}

File: app/desktop/src-tauri/src/services/vcfg.rs
//! VCFG / CFG 读取壳层 → core 纯逻辑。
//!
//! 对应原 `app/desktop/src/main/services/vcfg.ts` 的 I/O 部分：
//! 读取文件内容后调用 core 的 `parse_vdf` / `child` / `string_entries` /
//! `parse_cfg_convars` / `snapshot_to_cfg` / `normalize_cfg_value`。

use std::collections::HashMap;
use std::path::{Path, PathBuf};

use srp_cfg_core::{
    child, count_entries, parse_cfg_convars, parse_vdf, snapshot_to_cfg, string_entries,
    SnapshotToCfgOptions, VdfNode,
};

use crate::models::{VcfgSnapshot, VcfgStateSummary};

fn now_ms() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_millis() as u64)
        .unwrap_or(0)
}

fn read_vdf(path: &Path) -> Option<VdfNode> {
    std::fs::read_to_string(path).ok().and_then(|c| parse_vdf(&c))
}

fn string_map(node: Option<&VdfNode>) -> HashMap<String, String> {
    node.map(|n| string_entries(n).into_iter().collect())
        .unwrap_or_default()
}

/// 对应 TS `captureVcfgSnapshot`。
pub fn capture_vcfg_snapshot(user_cfg_path: &str) -> VcfgSnapshot {
    let dir = Path::new(user_cfg_path);
    let keys = read_vdf(&dir.join("cs2_user_keys_0_slot0.vcfg"));
    let user_convars = read_vdf(&dir.join("cs2_user_convars_0_slot0.vcfg"));
    let machine_convars = read_vdf(&dir.join("cs2_machine_convars.vcfg"));

    VcfgSnapshot {
        schema_version: 1,
        captured_at: now_ms(),
        user_cfg_path: user_cfg_path.to_string(),
        bindings: string_map(keys.as_ref().and_then(|k| child(k, &["config", "bindings"]))),
        analog_bindings: string_map(keys.as_ref().and_then(|k| child(k, &["config", "analogbindings"]))),
        user_convars: string_map(user_convars.as_ref().and_then(|k| child(k, &["config", "convars"]))),
        machine_convars: string_map(machine_convars.as_ref().and_then(|k| child(k, &["config", "convars"]))),
    }
}

/// 对应 TS `inspectVcfgState`。
pub fn inspect_vcfg_state(user_cfg_path: Option<&str>) -> VcfgStateSummary {
    let Some(user_cfg_path) = user_cfg_path else {
        return VcfgStateSummary::default();
    };
    let dir = Path::new(user_cfg_path);
    let keys_path = dir.join("cs2_user_keys_0_slot0.vcfg");
    let user_convars_path = dir.join("cs2_user_convars_0_slot0.vcfg");
    let machine_convars_path = dir.join("cs2_machine_convars.vcfg");

    let keys = read_vdf(&keys_path);
    let user_convars = read_vdf(&user_convars_path);
    let machine_convars = read_vdf(&machine_convars_path);
    let bindings = string_map(keys.as_ref().and_then(|k| child(k, &["config", "bindings"])));

    VcfgStateSummary {
        available: keys.is_some() || user_convars.is_some() || machine_convars.is_some(),
        bindings: bindings.len(),
        analog_bindings: count_entries(
            keys.as_ref().and_then(|k| child(k, &["config", "analogbindings"])).unwrap_or(&HashMap::new()),
        ),
        cloud_convars: count_entries(
            user_convars.as_ref().and_then(|k| child(k, &["config", "convars"])).unwrap_or(&HashMap::new()),
        ),
        machine_convars: count_entries(
            machine_convars.as_ref().and_then(|k| child(k, &["config", "convars"])).unwrap_or(&HashMap::new()),
        ),
        has_cloud_mirror: {
            let k = format!("{}.vcfg_lastclouded", keys_path.to_string_lossy());
            let u = format!("{}.vcfg_lastclouded", user_convars_path.to_string_lossy());
            Path::new(&k).exists() || Path::new(&u).exists()
        },
        has_video_config: dir.join("cs2_video.txt").exists(),
    }
}

/// 对应 TS `parseCfgConvars`：读文件 → core 解析（跳过 echo/exec/bind/注释）。
pub fn parse_cfg_convars_file(path: &Path) -> HashMap<String, String> {
    std::fs::read_to_string(path)
        .map(|c| parse_cfg_convars(&c).into_iter().collect())
        .unwrap_or_default()
}

/// 对应 TS `saveVcfgBaseline`：首次创建 baseline.json。
pub fn save_vcfg_baseline(
    user_cfg_path: &str,
    snapshot_root: &str,
    account_id: &str,
) -> (PathBuf, bool) {
    let account_dir = Path::new(snapshot_root).join(account_id);
    let baseline_path = account_dir.join("baseline.json");
    if baseline_path.exists() {
        return (baseline_path, false);
    }
    let _ = std::fs::create_dir_all(&account_dir);
    let snapshot = capture_vcfg_snapshot(user_cfg_path);
    let json = serde_json::to_string_pretty(&snapshot).unwrap_or_default();
    let _ = std::fs::write(&baseline_path, json);
    (baseline_path, true)
}

/// 快照 → CFG 文本（core 纯逻辑）。
pub fn snapshot_to_cfg_text(
    snapshot: &VcfgSnapshot,
    options: &SnapshotToCfgOptions,
    baseline: &HashMap<String, String>,
) -> String {
    let bindings: Vec<(String, String)> = snapshot.bindings.clone().into_iter().collect();
    let analog: Vec<(String, String)> = snapshot.analog_bindings.clone().into_iter().collect();
    let user: Vec<(String, String)> = snapshot.user_convars.clone().into_iter().collect();
    let machine: Vec<(String, String)> = snapshot.machine_convars.clone().into_iter().collect();
    snapshot_to_cfg(&bindings, &analog, &user, &machine, options, baseline)
}

